include/mbedtls/pkcs12.h

```diff
@@ -85,14 +87,16 @@ int mbedtls_pkcs12_pbe( mbedtls_asn1_buf *pbe_params, int mode,
  *                   integrity key.
  *
  * \param data       buffer to store the derived data in
- * \param datalen    length to fill
- * \param pwd        password to use (may be NULL if no password is used)
+ * \param datalen    length of buffer to fill
+ * \param pwd        Null terminated BMPString password to use (may be NULL if
+ *                   no password is used, but not if pwdlen is non-zero)
  * \param pwdlen     length of the password (may be 0)
```

Comment:
Actually, looking at our implementation, I don't think this function is correct when `pwdlen==0` or `saltlen==0`. This is unrelated to pointer nullability. Our code does:

```
    if( hlen <= 32 )
        v = 64;
    else
        v = 128;
…
    pkcs12_fill_buffer( salt_block, v, salt, saltlen );
    pkcs12_fill_buffer( pwd_block,  v, pwd,  pwdlen  );
…
        if( ( ret = mbedtls_md_update( &md_ctx, salt_block, v ) ) != 0 )
            goto exit;
        if( ( ret = mbedtls_md_update( &md_ctx, pwd_block, v ) ) != 0 )
            goto exit;
```
But [PKCS#12 §B.2](https://datatracker.ietf.org/doc/html/rfc7292#appendix-B.2) steps 2 and 3 specify that if the salt or password is empty, what gets hashed is an empty string, not `v` bytes.

`mbedtls_pkcs12_derivation` already errors out when `pwdlen > 64 || saltlen > 64`, both cases where the calculation quoted above would be incorrect because it would need to hash multiple blocks of `v` bytes. (64 could be raised to 128 when the hash is one with 128-byte blocks.) We could simply make it error out when `pwdlen == 0 || saltlen == 0` for the same reason.

Per RFC 128, `pwdlen` is at least 2 bytes (a UCS2 null byte). `saltlen` is an arbitrary byte string, and could in theory be empty. A salt is necessary for security when the password is actually a password, but not when using an empty password (`pwdlen==2`).

Erroring out would break backward compability, but it would eliminate a non-compliance with the standard. So I'm not sure what to do. At least in LTS branches, we should maintain backward compatibility, but document what's going on. In non-LTS branches, I would prefer to either error out or fix the functional bug. Fixing the invalid pointer dereference is urgent, fixing the functional bug isn't, so we don't need to fix the functional bug now, but if we decide not to make it an error, we should at least document the current behavior, either as something we guarantee or something we may fix in the future.

Reply:
Ah, ok. I took this documentation to mean that the password and salt buffers still had to be 'v' bytes long, but if the password or salt was empty, then so would the buffer be (hence the memset to zero)

Reply:
Hang on, though. How are we supposed to do step 6C from the documentation above if the password and salt blocks are not of length `v`

Reply:
Step 6 doesn't use the password and salt directly, only I. I is always a concatenation of v-bit blocks. The number of blocks can be 0; this only happens if both the salt and the password are empty. If the number of blocks is 0, step 6C doesn't actually modify I. This might hurt for the security analysis of the algorithm, but that's not really a concern when both the administrator and the user decided to have no effective security (0-length salt and empty password).